File: internal/realtime/sse.go
package realtime

import (
	"fmt"
	"net/http"
	"sync"

	"gold-bot/internal/domain"
)

type Hub struct {
	mu          sync.RWMutex
	subscribers map[chan domain.Event]struct{}
}

func NewHub() *Hub {
	return &Hub{
		subscribers: map[chan domain.Event]struct{}{},
	}
}

func (h *Hub) Subscribe() chan domain.Event {
	ch := make(chan domain.Event, 16)
	h.mu.Lock()
	h.subscribers[ch] = struct{}{}
	h.mu.Unlock()
	return ch
}

func (h *Hub) Unsubscribe(ch chan domain.Event) {
	h.mu.Lock()
	if _, ok := h.subscribers[ch]; ok {
		delete(h.subscribers, ch)
		close(ch)
	}
	h.mu.Unlock()
}

func (h *Hub) Publish(event domain.Event) {
	h.mu.RLock()
	defer h.mu.RUnlock()
	for ch := range h.subscribers {
		select {
		case ch <- event:
		default:
		}
	}
}

func (h *Hub) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")

	flusher, ok := w.(http.Flusher)
	if !ok {
		http.Error(w, "streaming unsupported", http.StatusInternalServerError)
		return
	}

	ch := h.Subscribe()
	defer h.Unsubscribe(ch)

	for {
		select {
		case <-r.Context().Done():
			return
		case event, ok := <-ch:
			if !ok {
				return
			}
			_, _ = fmt.Fprintf(w, "data: %s\n\n", event.JSON())
			flusher.Flush()
		}
	}
}

File: packages/observability/src/sse.ts
export function eventStreamHeaders(): Record<string, string> {
  return {
    'Content-Type': 'text/event-stream',
    'Cache-Control': 'no-cache',
    Connection: 'keep-alive'
  };
}

export function formatSseFrame(data: unknown): string {
  return `data: ${JSON.stringify(data)}\n\n`;
}

export type SseEvent = {
  event_id: string;
  event_type: string;
  account_id?: string;
  source: string;
  timestamp: string;
  payload: unknown;
};

export type SseSubscriber<T> = (event: T) => void;

export class SseHub<T = SseEvent> {
  private readonly subscribers = new Set<SseSubscriber<T>>();

  subscribe(subscriber: SseSubscriber<T>): () => void {
    this.subscribers.add(subscriber);
    return () => {
      this.subscribers.delete(subscriber);
    };
  }

  publish(event: T): void {
    for (const subscriber of this.subscribers) {
      subscriber(event);
    }
  }

  subscriberCount(): number {
    return this.subscribers.size;
  }
}

export function createSseHub<T = SseEvent>(): SseHub<T> {
  return new SseHub<T>();
}
